### PR TITLE
Remove last use of mem::uninitialized from std::io::util

### LINKED ISSUE #62397: Get rid of uses of mem::uninitialized
https://github.com/rust-lang/rust/commit/007d87f1719e2fcf2ff36aef8b6dc866fa276386 added a bunch of `allow(deprecated)` because there are still a few places where we use `mem::uninitialized`. These should all be fixed, to remove the `allow` (and hopefully prevent other deprecated methods from being used there).

* [x] SGX
* [x] libterm
* [x] librustc_codegen_llvm
* [x] std::io::util
* [ ] CloudAbi layout tests (`src/libstd/sys/cloudabi/abi/cloudabi.rs`)
* [ ] rustc tests

Cc @EdSchouten  for CloudAbi, @jethrogb for SGX.

### PR DESCRIPTION
**Pull request by nathanwhit**
Addresses #62397 for std::io::util

### CODE REVIEW

**Review by dtolnay — APPROVED** [review]
Looks good to me.

**Comment by dtolnay** [comment]
Adding another reviewer, since this PR involves unsafe code.

r? @Amanieu

**Comment by Amanieu** [comment]
My impression is that the proper way of dealing with uninitialized arrays is to make the *element* `MaybeUninit`:

```rust
        let mut buf: [mem::MaybeUninit<u8>; super::DEFAULT_BUF_SIZE] = [mem::MaybeUninit::uninit(); super::DEFAULT_BUF_SIZE];
        let slice = slice::from_raw_parts_mut(MaybeUninit::first_ptr_mut(&mut buf), super::DEFAULT_BUF_SIZE);
        reader.initializer().initialize(slice);
```
Not that it really matters in this case since we're creating a reference to slice of uninitialized data...

**Comment by nathanwhit** [comment]
> My impression is that the proper way of dealing with uninitialized arrays is to make the _element_ `MaybeUninit`:
> 
> ```rust
>         let mut buf: [mem::MaybeUninit<u8>; super::DEFAULT_BUF_SIZE] = [mem::MaybeUninit::uninit(); super::DEFAULT_BUF_SIZE];
>         let slice = slice::from_raw_parts_mut(MaybeUninit::first_ptr_mut(&mut buf), super::DEFAULT_BUF_SIZE);
>         reader.initializer().initialize(slice);
> ```
> 
> Not that it really matters in this case since we're creating a reference to slice of uninitialized data...

That was my first inclination as well, but it looks making a slice from a MaybeUninit is an unstable library feature, and I think a handful of other methods would have to change in order for it to type check with that approach. If you think it's worth it, I can definitely go that route though

**Comment by Amanieu** [comment]
It's probably not worth the trouble in this case. The slice APIs for `MaybeUninit` are still very primitive.

File: src/libstd/io/util.rs [hunk 1]
```diff
@@ -44,10 +44,12 @@ pub fn copy<R: ?Sized, W: ?Sized>(reader: &mut R, writer: &mut W) -> io::Result<
     where R: Read, W: Write
 {
     let mut buf = unsafe {
-        #[allow(deprecated)]
-        let mut buf: [u8; super::DEFAULT_BUF_SIZE] = mem::uninitialized();
-        reader.initializer().initialize(&mut buf);
-        buf
+        // This is still technically undefined behavior due to creating a reference
```
Comment (lzutao) [inline src/libstd/io/util.rs]: Maybe add a FIXME here as a reminder?

Comment (nathanwhit) [reply src/libstd/io/util.rs]: Sounds reasonable to me, I'll add that in

**Comment by RalfJung** [comment]
Looks like we collided mid-air here.^^
https://github.com/rust-lang/rust/pull/62746

**Comment by RalfJung** [comment]
But i think we can do slightly better than what this PR does, by not ever having uninitialized by-value integers. So I am going to propose we still merge my PR.

File: src/libstd/io/util.rs [hunk 2]
```diff
@@ -44,10 +44,15 @@ pub fn copy<R: ?Sized, W: ?Sized>(reader: &mut R, writer: &mut W) -> io::Result<
...
+        // to uninitialized data, but within libstd we can rely on more guarantees
+        // than if this code were in an external lib
```
Comment (RalfJung) [inline src/libstd/io/util.rs]: This doesn't just create a *reference* to uninitialized data (so the comment is not entirely correct). It actually creates (in the final `assume_init`) *uninitialized (integer) values* (not behind a reference).

**Pull request merged**